metricproviders/kayenta/kayenta.go

```diff
@@ -0,0 +1,257 @@
+package kayenta
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+	templateutil "github.com/argoproj/argo-rollouts/utils/template"
+)
+
+const (
+	//ProviderType indicates the provider is kayenta
+	ProviderType   = "Kayenta"
+	scoreURLFormat = `%s/canary/%s`
+
+	configIdLookupURLFormat = `%s/canaryConfig?application=%s&configurationAccountName=%s`
+
+	jobURLFormat = `%s/canary/%s?application=%s&metricsAccountName=%s&configurationAccountName=%s&storageAccountName=%s`
+
+	jobPayloadFormat = `
+							{
+								"scopes": {
+										%s
+								},
+                                "thresholds" : {
+                                    "pass": %d,
+                                    "marginal": %d
+                                }
+                            }`
+
+	resumeDelay          time.Duration = 15 * time.Second
+	httpConnectionTimout time.Duration = 15 * time.Second
+	scopeFormat                        = `"%s":{"controlScope": %s, "experimentScope": %s}`
+)
+
+type Provider struct {
+	logCtx log.Entry
+	client http.Client
+}
+
+type canaryConfig struct {
+	Id                  string
+	Name                string
+	UpdatedTimestamp    int
+	UpdatedTimestampIso string
+	Applications        []string
+}
+
+// Type incidates provider is a kayenta provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+func getCanaryConfigId(metric v1alpha1.Metric, p *Provider) (string, error) {
+	canaryConfigId := ""
+	configIdLookupURL := fmt.Sprintf(configIdLookupURLFormat, metric.Provider.Kayenta.Address, metric.Provider.Kayenta.Application, metric.Provider.Kayenta.StorageAccountName)
+
+	response, err := p.client.Get(configIdLookupURL)
+	if err != nil || response.Body == nil || response.StatusCode != 200 {
+		if err == nil {
+			err = errors.New("Invalid Response")
+		}
+		return canaryConfigId, err
```

Comment:
Instead of using canaryConfigId variable, can we return an empty string here? This makes it more clear that we are in an error state and that the canaryConfigID is not set. Otherwise, we have to scroll through the function to confirm that the string is not set.

Reply:
No. that will not work if there is no match with the if. Then there will be no return.